### ISSUE TITLE
Lambda Layer is crashing because current_root_span is returning None

### ISSUE THREAD

**Issue by amersaw** [DataDog/dd-trace-py#5220]
### Summary of problem
Today we start facing crashes on Lambda app that was using the [datadog-lambda-python](https://github.com/DataDog/datadog-lambda-python) extension. I checked the logs and found that crash caused by `ddtrace`. 

### Which version of dd-trace-py are you using?
1.9.0
The `datadog-lambda-python` package is using  the latest version of ddtrace (<2) 

### Which version of pip are you using?
pip 22.3.1

### Which libraries and their versions are you using?
```
datadog==0.44.0
datadog-lambda==4.68.0
ddsketch==2.0.4
ddtrace==1.9.0
```

### How can we reproduce your problem?
- Deploy a lambda app as described in [the following official documentation](https://docs.datadoghq.com/serverless/installation/python/?tab=containerimage)
- Couldn't determine when exactly the app will crash, but about 30% of requests are crashing where other requests working as expected.
- When the request crash I get the output log mentioned below.

### What is the result that you get?
The whole lambda request fails with the following Cloudwatch logs:

```
START RequestId: bac50653-2b72-49c7-86e7-ad6695169298 Version: $LATEST
Traceback (most recent call last):
File "/var/runtime/bootstrap", line 9, in <module>
main()
File "/var/runtime/bootstrap.py", line 346, in main
event_request = lambda_runtime_client.wait_next_invocation()
File "/var/runtime/lambda_runtime_client.py", line 57, in wait_next_invocation
response = self.runtime_connection.getresponse()
File "/var/lang/lib/python3.7/http/client.py", line 1373, in getresponse
response.begin()
File "/var/lang/lib/python3.7/http/client.py", line 319, in begin
version, status, reason = self._read_status()
File "/var/lang/lib/python3.7/http/client.py", line 280, in _read_status
line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
File "/var/lang/lib/python3.7/socket.py", line 589, in readinto
return self._sock.recv_into(b)
File "/var/task/ddtrace/contrib/aws_lambda/patch.py", line 47, in wrap_signals
f(*args, **kwargs)
File "/var/task/ddtrace/contrib/aws_lambda/patch.py", line 25, in _crash_flush
root_span.error = 1
AttributeError: 'NoneType' object has no attribute 'error'
RequestId: bac50653-2b72-49c7-86e7-ad6695169298 Error: Runtime exited with error: exit status 1 Runtime.ExitError
END RequestId: bac50653-2b72-49c7-86e7-ad6695169298 
```

### What is the result that you expected?
Such as error shouldn't occur.

**Comment by amersaw** [comment]
Note that here we don't check that `root_span` is not None  [`patch.py:18`](https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/contrib/aws_lambda/patch.py#L18)
```
def _crash_flush(_, __):
    root_span = tracer.current_root_span()
    root_span.error = 1
    ...
    ...

```

**Comment by duncanista** [comment]
Hey @amersaw,

Thanks for bringing this to our attention. I'll be working shortly on it. 
Could you provide some more details about your Lambda function? I'm curious to understand why this is happening. The only thing that I can think off is that your Lambda function might be running out of time, but all of the spans have been closed for some reason. Hence lacking a root span.

This function (`_crash_flush`) will never run unless your function reaches a point which might end up into a timeout. Please, let me know if you can provide more details!

- jordan

**Comment by amersaw** [comment]
Hi @duncanista 

Thanks for your feedback, note the following:
* Our Datadog integration is yet in the early stages, so we only added the lambda layer and changed the lambda handler to use Datadog's one. which means within our code there's no modification being done to any spans.

* I checked the Cloudwatch logs and found the total execution time of crashing requests is about 2s (Where normally the function is taking 5s maximum), the function timeout set to be 15s

* Unfortunately this affected our prod environment because the `datadog-lambda` package is always using the latest version of `ddtrace`.

**Comment by duncanista** [comment]
Hey @amersaw 

Are you using the python `signal` module? That's the only thing that comes to my mind, considering all the information you provided me. 

This PR should be merged soon, will still follow up in this thread. 
If you have any other concerns, I'd suggest raising a ticket with Datadog so we can explore more in depth with your account. One engineer, or myself, can look directly into the problem one escalated, this would help us to find more regressions in this specific scenario.

Thanks!

**Comment by JJSphar** [comment]
I was able to work around this crash by upping the amount of provisioned memory for my function.

The function wasn't close to timing out (20s out of 60s limit) but it was running into high memory usage. Bumping the amount of memory seems to help it run faster and avoid this trap.

**Comment by duncanista** [comment]
The PR for the fix is now merged. I won't close this since I want to make sure we can provide more insights on why this could be happening.

**Comment by amer-idwise** [comment]
Here are few observations I noticed about the crashing function:
* There are ~20 functions deployed from the exact same code base, but only 1 function was crashing.
* That crashing function has less resources assigned (20-sec timeout + 512 MB Memory) than the other not-crashing functions  (500-sec timeout / 3GB Memory).
* The max memory consumption for the function was 165 MB (Including the crashing requests).
![image](https://user-images.githubusercontent.com/117313207/222563220-f6d68ba8-5655-4862-9d2e-6ec4621bbe58.png)
* No we're not using the `signal` module in any part of our code.
* We use multi-threading (Main thread + 1 thread).

Thanks @JJSphar for noting the memory-related issue, have you noticed that the request is using more RAM than the assigned RAM, or it's just almost hitting the limit? 

@duncanista For now I'm using the older `1.8` version of `ddtrace` package, but still no clear reason why the process would crash even though it's not hitting the assigned timeout (~ 3s out of 20s) nor the assigned memory limits (~70MB out of 512MB).

**Comment by amersaw** [comment]
Here are few observations I noticed about the crashing function:
* There are ~20 functions deployed from the exact same code base, but only 1 function was crashing.
* That crashing function has less resources assigned (20-sec timeout + 512 MB Memory) than the other not-crashing functions  (500-sec timeout / 3GB Memory).
* The max memory consumption for the function was 165 MB (Including the crashing requests).
![image](https://user-images.githubusercontent.com/117313207/222563220-f6d68ba8-5655-4862-9d2e-6ec4621bbe58.png)
* No we're not using the `signal` module in any part of our code.
* We use multi-threading (Main thread + 1 thread).

Thanks @JJSphar for noting the memory-related issue, have you noticed that the request is using more RAM than the assigned RAM, or it's just almost hitting the limit? 

@duncanista For now I'm using the older `1.8` version of `ddtrace` package, but still no clear reason why the process would crash even though it's not hitting the assigned timeout (~ 3s out of 20s) nor the assigned memory limits (~70MB out of 512MB).

**Comment by luisabtrace** [comment]
This started happening randomly for us yesterday. We did not change the versions of any datadog dependencies. `npx @datadog/datadog-ci lambda uninstrument` seems to get rid of the errors, but we lose logging.

**Comment by duncanista** [comment]
I've added a backport to fix this issue in `1.9`. I'll keep researching to understand this problem. I haven't been successful reproducing this issue. 

A small test case would be really useful, either provided via a support ticket or here. Thanks everyone!

**Comment by duncanista** [comment]
Shipped in https://github.com/DataDog/dd-trace-py/releases/tag/v1.9.2
